gpflow/optimizers/mcmc.py

```diff
@@ -84,6 +84,14 @@ def _target_log_prob_fn_closure(*variables):
             with tf.GradientTape(watch_accessed_variables=False) as tape:
                 tape.watch(variables_list)
                 log_prob = self._target_log_prob_fn()
+                # Now need to correct for the fact that the prob fn is evaluated on the
+                # constrained space while we wish to evaluate it in the unconstrained space
+                for param in self._model_parameters:
+                    if isinstance(param, Parameter) and param.transform is not None:
+                        x = param.unconstrained_variable
+                        log_det_jacobian = param.transform.forward_log_det_jacobian(x,
```

Comment:
`forward_log_det_jacobian` may not be defined if it's a non-injective bijector. Why not instead subtract the `inverse_log_det_jacobian` to cancel the (wrong, in this case) correction in `param.log_prior()`?

Reply:
Could do! I haven't thought about the case of non-injective bijectors - it might be safest to assert that the transform must be injective if a prior is specified.

Reply:
So what would be the (dis)advantages of having this code as follows instead:
```python
                for param in self._model_parameters:
                    if isinstance(param, Parameter) and param.transform is not None:
                        y = param.read_value()
                        log_det_jacobian = param.transform.inverse_log_det_jacobian(y,
                                                                                    y.shape.ndims)
                        log_prob -= tf.reduce_sum(log_det_jacobian)
```
exactly mirroring the behaviour in `param.log_prior()` ?

Reply:
Ah because in param.log_prior() the inverse_log_det_jacobian is only applied when the prior is defined on the unconstrained variable. If the prior is defined on the constrained variable, then we aren't trying to cancel a term, we're applying a correction for the first time.

Reply:
And that correction is (I think) potentially pathological if the transform isn't bijective.